pybandits/simulator.py

```diff
@@ -0,0 +1,592 @@
+
+        Parameters
+        ----------
+        actions : List[ActionId]
+            The actions selected by the multi-armed bandit model.
+        metadata : Dict[str, List]
+            The metadata for the selected actions.
+
+        Returns
+        -------
+        reward : List[BinaryReward]
+            A list of binary rewards.
+        """
+        pass
+
+    @abstractmethod
+    def _get_batch_step_kwargs_and_metadata(
+        self, batch_index: int
+    ) -> Tuple[Dict[str, Union[int, np.ndarray]], Dict[str, np.ndarray], Dict[str, List]]:
+        """
+        Extract kwargs required for the MAB's update and predict functionality,
+        as well as metadata for sample association.
+
+        Parameters
+        ----------
+        batch_index : int
+            The index of the batch.
+
+        Returns
+        -------
+        predict_kwargs : Dict[str, Union[int, np.ndarray]]
+            Dictionary containing the keyword arguments for the batch used in mab.predict.
+        update_kwargs : Dict[str, Any]
+            Dictionary containing the keyword arguments for the batch used in mab.update.
+        metadata : Dict[str, List]
+            Dictionary containing the association information for the batch.
+        """
+        pass
+
+    @abstractmethod
+    def _finalize_step(self, batch_results: pd.DataFrame) -> pd.DataFrame:
+        """
+        Finalize the step by adding additional information to the batch results.
+
+        Parameters
+        ----------
+        batch_results : pd.DataFrame
+            raw batch results
+
+        Returns
+        -------
+        batch_results : pd.DataFrame
+            batch results with added columns
+        """
+        pass
+
+    @abstractmethod
+    def _finalize_results(self):
+        """
+        Finalize the simulation process. It can be used to add additional information to the results.
+
+        Returns
+        -------
+        None
+        """
+        pass
+
+    @cached_property
+    def _actions(self) -> List[ActionId]:
+        """
+        Get the list of actions.
+
+        Returns
+        -------
+        List[ActionId]
+            The list of actions
+        """
+        return sorted(list(self.mab.actions.keys()))
+
+    @cached_property
+    def _cumulative_actions_cols(self) -> List[str]:
+        """
+        Get the list of cumulative actions columns.
+
+        Returns
+        -------
+        : List[str]
+            The list of cumulative actions columns
+        """
+        return [f"{self._cumulative_col_prefix}_{action}" for action in self._actions]
+
+    @property
+    def _colors(self) -> Palette:
+        """
+        Get the palette of colors.
+
+        Returns
+        -------
+        : Palette
+            Palette of colors
+        """
+        n_actions = len(self._actions)
+        return Category10[max(n_actions, 3)] if n_actions <= 10 else Turbo256
```

Comment:
not a big fan of hardcoded numbers in the middle of the code, can you please extract them with a meaningful name?
COLORS_SOME_NAME = 3
COLORS_SOMEOTHER_NAME = 10

Reply:
I agree, but I implemented it a little different than your proposal. Take alook :-)

Reply:
very good!